controllers/helmchartproxy/helmchartproxy_controller.go

```diff
@@ -197,6 +209,102 @@ func (r *HelmChartProxyReconciler) reconcileNormal(ctx context.Context, helmChar
 		}
 	}
 
+	// RolloutStepSize is defined; check if count of HelmReleaseProxies matches
+	// the count of matching clusters.
+	if helmChartProxy.Spec.RolloutStepSize != nil {
+		if len(clusters) != len(helmReleaseProxies) {
+			// Set HelmReleaseProxiesRolloutCompletedCondition to false as
+			// HelmReleaseProxies are being rolled out.
+			conditions.MarkFalse(
+				helmChartProxy,
+				addonsv1alpha1.HelmReleaseProxiesRolloutCompletedCondition,
+				addonsv1alpha1.HelmReleaseProxiesRolloutNotCompleteReason,
+				clusterv1.ConditionSeverityInfo,
+				"%d Helm release proxies not yet rolled out",
+				len(clusters)-len(helmReleaseProxies),
+			)
+
+			// Identifies clusters by their NamespacedName and gathers their
+			// helmReleaseProxyRolloutMeta.
+			clusterRolloutMeta := map[string]*helmReleaseProxyRolloutMeta{}
+			for _, cls := range clusters {
+				clusterRolloutMeta[types.NamespacedName{Namespace: cls.Namespace, Name: cls.Name}.String()] = &helmReleaseProxyRolloutMeta{cluster: cls}
+			}
+			for _, hrp := range helmReleaseProxies {
+				hrpClsRef := hrp.Spec.ClusterRef
+				rltMeta := clusterRolloutMeta[types.NamespacedName{Namespace: hrpClsRef.Namespace, Name: hrpClsRef.Name}.String()]
+				rltMeta.hrpExists = true
+			}
+
+			hrpReadyCond := conditions.Get(helmChartProxy, addonsv1alpha1.HelmReleaseProxiesReadyCondition)
+			// If HelmReleaseProxiesReadyCondition is false, reconcile existing
+			// HelmReleaseProxies and exit.
+			if hrpReadyCond != nil && (hrpReadyCond.Status == corev1.ConditionFalse) {
+				for _, hrpRltMeta := range clusterRolloutMeta {
+					if hrpRltMeta.hrpExists {
+						// Don't reconcile if the Cluster is being deleted
+						if !hrpRltMeta.cluster.DeletionTimestamp.IsZero() {
+							continue
+						}
+
+						err := r.reconcileForCluster(ctx, helmChartProxy, hrpRltMeta.cluster)
+						if err != nil {
+							return err
+						}
+					}
+				}
+
+				return nil
+			}
+
+			// HelmReleaseProxyReadyCondition is True; continue with reconciling the
+			// next batch of HelmReleaseProxies.
+			count := 0
+			stepSize, err := intstr.GetScaledValueFromIntOrPercent(helmChartProxy.Spec.RolloutStepSize, len(clusters), true)
+			if err != nil {
+				return err
+			}
+			for _, hrpRltMeta := range clusterRolloutMeta {
+				// The next batch of helmReleaseProxies have been reconciled.
+				if count >= stepSize {
+					return nil
+				}
+
+				// Skip reconciling the cluster if its HelmReleaseProxy already exists.
+				if hrpRltMeta.hrpExists {
+					continue
+				}
+
+				// Don't reconcile if the Cluster is being deleted
+				if !hrpRltMeta.cluster.DeletionTimestamp.IsZero() {
+					continue
+				}
+
+				err := r.reconcileForCluster(ctx, helmChartProxy, hrpRltMeta.cluster)
+				if err != nil {
+					return err
+				}
```

Comment:
There are very similar code to the code in lines 244-255, except condition in hrpRltMeta.hrpExists. You can wrap it with function (probably anonymous) and use it here and upper.

Reply:
Moved cluster deletion timestamp check to within the reconcileForCluster() to make those sections look cleaner. Let me know if this works for you.

Reply:
Ok, good